### PR TITLE
Use pipgraphfragmentmanager for threading instead of frontendmanager

### PR DESCRIPTION
**Pull request by erickulcyk**
Appears to fix hangs in evaluation.

### CODE REVIEW

**Comment by ljcollins25** [comment]
>     public class PipGraphFragmentSerializer

I think this class should be left synchronous, and the handleDeserializedPip should handle queuing tasks.

---
Refers to: Public/Src/Pips/Dll/Operations/PipGraphFragmentSerializer.cs:20 in e58ecfe. [](commit_id = e58ecfea1817c13c62d3b1137acb0099dd714416, deletion_comment = False)

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 1]
```diff
@@ -35,6 +39,10 @@ public class PipGraphFragmentManager : IPipGraphFragmentManager
         private readonly ConcurrentBigMap<long, PipId> m_semiStableHashToPipId = new ConcurrentBigMap<long, PipId>();
         private readonly ConcurrentBigMap<long, DirectoryArtifact> m_semiStableHashToDirectory = new ConcurrentBigMap<long, DirectoryArtifact>();
 
+        private readonly TaskFactory m_taskFactory = new TaskFactory(new DedicatedThreadsTaskScheduler(Environment.ProcessorCount * 2, "PipGraphFragmentManager"));
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >Environment.ProcessorCount * 2 [](start = 103, length = 30)

Is double actually better?

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 2]
```diff
@@ -48,24 +56,44 @@ public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContex
         /// <summary>
         /// Adds a single pip graph fragment to the graph.
         /// </summary>
-        public bool AddFragmentFileToGraph(AbsolutePath filePath, string description)
+        public bool AddFragmentFileToGraph(AbsolutePath filePath, string description, IEnumerable<AbsolutePath> dependencies)
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >IEnumerable<AbsolutePath> dependencies [](start = 86, length = 38)

First ensure all dependencies are in the task map.

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 3]
```diff
@@ -48,24 +56,44 @@ public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContex
...
-            {
-                Logger.Log.ExceptionOnDeserializingPipGraphFragment(m_loggingContext, filePath.ToString(m_context.PathTable), e.ToString());
-                return false;
-            }
+                try
+                {
+                    var result = await deserializer.DeserializeAsync(
+                        filePath,
+                        m_taskFactory,
+                        (fragmentContext, provenance, pipId, pip) => AddPipToGraph(fragmentContext, provenance, pipId, pip),
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >AddPipToGraph [](start = 69, length = 13)

Capture task to add to pip graph in a list and then do a when all after deserialization (not async) completes.

**Comment by ljcollins25** [comment]
>         /// Adds a single pip graph fragment to the graph.

Can we add to the comment to indicate that this is only for testing purposes.

---
Refers to: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs:100 in e58ecfe. [](commit_id = e58ecfea1817c13c62d3b1137acb0099dd714416, deletion_comment = False)

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 4]
```diff
@@ -76,11 +104,12 @@ public bool AddFragmentFileToGraph(Stream stream, string description)
...
             {
-                var result = deserializer.Deserialize(
+                var result = deserializer.DeserializeAsync(
                     stream,
+                    m_taskFactory,
                     (fragmentContext, provenance, pipId, pip) => AddPipToGraph(fragmentContext, provenance, pipId, pip),
-                    description);
-                Logger.Log.DeserializationStatsPipGraphFragment(m_loggingContext, deserializer.FragmentDescription, deserializer.Stats.ToString());
+                    description).Result;
+                // Logger.Log.DeserializationStatsPipGraphFragment(m_loggingContext, deserializer.FragmentDescription, deserializer.Stats.ToString());
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >DeserializationStatsPipGraphFragment [](start = 30, length = 36)

Why is logging commented out?

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 5]
```diff
@@ -48,24 +56,44 @@ public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContex
...
-                return false;
-            }
+                try
+                {
+                    var result = await deserializer.DeserializeAsync(
+                        filePath,
+                        m_taskFactory,
+                        (fragmentContext, provenance, pipId, pip) => AddPipToGraph(fragmentContext, provenance, pipId, pip),
+                        description);
+                    // Logger.Log.DeserializationStatsPipGraphFragment(m_loggingContext, deserializer.FragmentDescription, deserializer.Stats.ToString());
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >DeserializationStatsPipGraphFragment [](start = 34, length = 36)

Why is logging commented out?

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 6]
```diff
@@ -48,24 +56,44 @@ public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContex
...
+                    // m_taskMap.TryRemove(filePath, out var _);
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >TryRemove [](start = 33, length = 9)

Remove this line.

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 7]
```diff
@@ -35,51 +40,97 @@ public class PipGraphFragmentManager : IPipGraphFragmentManager
         private readonly ConcurrentBigMap<long, PipId> m_semiStableHashToPipId = new ConcurrentBigMap<long, PipId>();
         private readonly ConcurrentBigMap<long, DirectoryArtifact> m_semiStableHashToDirectory = new ConcurrentBigMap<long, DirectoryArtifact>();
 
+        private readonly TaskFactory m_taskFactory = new TaskFactory(new DedicatedThreadsTaskScheduler(Environment.ProcessorCount, "Add Pip to Graph"));
+
+        private readonly ConcurrentBigMap<AbsolutePath, (PipGraphFragmentSerializer, Task<bool>)> m_taskMap = new ConcurrentBigMap<AbsolutePath, (PipGraphFragmentSerializer, Task<bool>)>();
```
Comment (SergeyTeplyakov) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: very nit: I would prefer to use names here, but this is up to you. And it seems that unnamed tuples are used everywhere anyways.

**Pull request merged**